lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp

```diff
@@ -214,4 +225,122 @@ void LLDBMemoryReader::popLocalBuffer() {
+  }
+
+#ifndef NDEBUG
+  static std::initializer_list<uint64_t> used_swift_masks = {
+      SWIFT_ABI_ARM_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_ARM_OBJC_WEAK_REFERENCE_MARKER_VALUE,
+      SWIFT_ABI_ARM_IS_OBJC_BIT,
+      SWIFT_ABI_X86_64_SWIFT_SPARE_BITS_MASK,
+      SWIFT_ABI_X86_64_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_X86_64_IS_OBJC_BIT,
+      SWIFT_ABI_X86_64_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_WEAK_REFERENCE_MARKER_VALUE,
+      SWIFT_ABI_ARM64_SWIFT_SPARE_BITS_MASK,
+      SWIFT_ABI_ARM64_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_ARM64_IS_OBJC_BIT,
+      SWIFT_ABI_ARM64_OBJC_WEAK_REFERENCE_MARKER_MASK};
+
+  for (auto swift_mask : used_swift_masks) {
+    assert((next_available_address & swift_mask) != swift_mask &&
+           "LLDB artificial address clashes with a swift mask!");
+  }
+#endif
+
+  SectionList *section_list = module->GetObjectFile()->GetSectionList();
+
+  auto section_list_size = section_list->GetSize();
+  if (section_list_size == 0)
+    return {};
+
+  auto last_section =
+      section_list->GetSectionAtIndex(section_list->GetSize() - 1);
+  // The virtual file address + the size of last section gives us the total size
+  // of this image in memory.
+  uint64_t size = last_section->GetFileAddress() + last_section->GetByteSize();
+  auto next_available_address_end = next_available_address + size;
+
+  // Keep the end addresses pointer aligned.
+  next_available_address_end = llvm::alignTo(next_available_address_end, 8);
+  m_range_module_map.emplace_back(next_available_address_end, module);
+  return {{next_available_address, next_available_address_end}};
+}
+
+llvm::Optional<Address>
+LLDBMemoryReader::resolveFromArtificialToLLDBAddress(uint64_t address) const {
+  Log *log(GetLogIfAllCategoriesSet(LIBLLDB_LOG_TYPES));
+
+  if (!m_process.GetTarget().GetSwiftReadMetadataFromFileCache())
+    return {};
+
+  // If the address contains our mask, this is an image we registered.
+  if (!(address & ARTIFICIAL_ADDRESS_MASK))
+    return {};
+
+  // Dummy pair with the address we're looking for.
+  auto comparison_pair = std::make_pair(address, ModuleSP());
+
+  // Explicitly compare only the addresses, never the modules in the pairs.
+  auto pair_iterator = std::lower_bound(
+      m_range_module_map.begin(), m_range_module_map.end(), comparison_pair,
+      [](auto &a, auto &b) { return a.first < b.first; });
+
+  if (pair_iterator != m_range_module_map.end()) {
+    uint64_t masked_file_address = pair_iterator->first;
+    ModuleSP module = pair_iterator->second;
+    uint64_t file_address;
+    if (pair_iterator == m_range_module_map.begin())
+      // Since this is the first registered module,
+      // just unsetting the mask will give the virtual file address.
+      file_address = address & ~ARTIFICIAL_ADDRESS_MASK;
```

Comment:
There needs to be a check:
```
if (file_address < _pair_iterator.first) {
  LLDB_LOG(log, "[MemoryReader] Address is tagged, but doesn't exist in m_range_module_map");
  return {}
}
```

Reply:
Similarly we could also store the last valid virtual address and have a sanity upper bounds check as well

Reply:
In the map, we store the end address, not start address, of each image, `std::lower_bound`: "Returns an iterator pointing to the first element in the range [first, last) that is not less than (i.e. greater or equal to) value, or last if no such element is found." 

Let's say that the fist image goes from 0x200...00 to 0x200..10, and we're looking for a tagged address 0x200..05. We'll end in this first case, and because we know the address is tagged, we know for sure it's in range.

For the upper bound, we already check `pair_iterator != m_range_module_map.end()`, `m_range_module_map.end()` is what will be return if the address we're looking for is larger than anything in the map.